pandas/core/groupby/groupby.py

```diff
@@ -1707,7 +1708,12 @@ def nth(self,
             if not self.as_index:
                 return out
 
-            out.index = self.grouper.result_index[ids[mask]]
+            result_index = self.grouper.result_index
+            out.index = result_index[ids[mask]]
+
+            if not self.observed and isinstance(
```

Comment:
I was hoping to get this into `_wrap_aggregated_output` (which is what wraps most of the other methods) but that method doesn't appear to do any reindexing so would have to mess with that to make it work. Figured this as a one-off was easier though ideally would have everything consolidated...

Reply:
you just need a small change I think in ``result_index`. This is very similar to what we do with ``all_grouper``.

Reply:
IIUC I think `result_index` is fine and handles properly; the problem here is specific to `nth` where the mask on the preceding lines essentially ignores the work that `result_index` does to properly handle NA categorical data, hence the reindex

Reply:
I am not convinced here; this looks like just a bandaid to me and there is an underlying issue that needs fixing. also what about dropna=True?

Reply:
`dropna='all'` seems to have a separate set of issues. Here's behavior on master:

```python
>>> grp.nth(0, dropna='all')
cat
a    1.0
b    2.0
c    3.0
Name: ser, dtype: float64
```
Which is not correct since `b` and `c` are never associated with those values. 

Also trying to mix that with observed doesn't work:

```python-traceback
>>> grp = df.groupby('cat', observed=True)['ser']
>>> grp.nth(0, dropna='all')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/williamayd/clones/pandas/pandas/core/groupby/groupby.py", line 1773, in nth
    result.index = self.grouper.result_index
  File "/Users/williamayd/clones/pandas/pandas/core/generic.py", line 5144, in __setattr__
    return object.__setattr__(self, name, value)
  File "pandas/_libs/properties.pyx", line 67, in pandas._libs.properties.AxisProperty.__set__
    obj._set_axis(self.axis, value)
  File "/Users/williamayd/clones/pandas/pandas/core/series.py", line 381, in _set_axis
    self._data.set_axis(axis, labels)
  File "/Users/williamayd/clones/pandas/pandas/core/internals/managers.py", line 155, in set_axis
    'values have {new} elements'.format(old=old_len, new=new_len))
ValueError: Length mismatch: Expected axis has 3 elements, new values have 1 elements
```

So both weird though I'm not sure how we want to handle the combinations of observed and dropna; will open a separate issue

Reply:
See #26454